---
output: html_notebook
title: 'GeoMx report'
params:
  pkc_summary_file: ''
  ntc_summary_file: ''
  qc_segments_summary_file: ''
  plot_list_qc_segments_file: ''
  plot_list_qc_probes_file: ''
  qc_probes_summary_file: ''
  goi_detection_rate_table_file: ''
  plot_list_normalization_file: ''
  plot_list_16s_file: ''
  plot_list_unsupervised_clustering_grid_file: ''
  plot_list_16s_score_file: ''
  plot_list_cv_heatmap_file: ''
  plot_list_diff_exprs_grid_file: ''
  plot_list_pathway_analysis_grid_file: ''
  plot_list_imm_decon_file: ''
  plot_list_tcr_analysis_grid_file: ''
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE, # Hides code
  warning = FALSE,
  message = FALSE,
  error = FALSE
  )

# Getting arguments from the command line:
# https://stackoverflow.com/questions/49904943/run-rmarkdown-with-arguments-on-the-command-line/76567366#76567366
# https://bookdown.org/yihui/rmarkdown-cookbook/parameterized-reports.html#parameterized-reports

# Dynamically building Rmd files:
# https://stackoverflow.com/questions/62754099/dynamically-building-rmarkdown
# https://stackoverflow.com/questions/21729415/generate-dynamic-r-markdown-blocks

# Reading in external scripts into a chunk:
# https://bookdown.org/yihui/rmarkdown-cookbook/option-code.html

# How to do tabbed sections: 
# https://epirhandbook.com/new_pages/rmarkdown.html

# Embedding Shiny apps:
# https://bookdown.org/yihui/rmarkdown/shiny-embedded.html
```

# Study design
```{r}
if(file.exists(params$pkc_summary_file)) {
  pkc_summary_df <- readRDS(params$pkc_summary_file)
  knitr::kable(pkc_summary_df)
}

```

# QC - segments
```{r}
if(file.exists(params$qc_segments_summary_file)) {
  qc_segments_summary_df <- readRDS(params$qc_segments_summary_file)
  knitr::kable(qc_segments_summary_df)
}

if(file.exists(params$ntc_summary_file)) {
  ntc_summary_df <- readRDS(params$ntc_summary_file)
  knitr::kable(ntc_summary_df)
}
```

```{r}
if(file.exists(params$plot_list_qc_segments_file)) {
  plot_list_qc_segments <- readRDS(params$plot_list_qc_segments_file)
  for(item in plot_list_qc_segments) {
    suppressMessages(capture.output(print(item)))
  }
}
```

# QC - probes and targets
```{r}
if(file.exists(params$qc_probes_summary_file)) {
  qc_probes_summary <- readRDS(params$qc_probes_summary_file)
  knitr::kable(qc_probes_summary)
}
```

```{r}
if(file.exists(params$goi_detection_rate_table_file)) {
  goi_detection_rate_table <- readRDS(params$goi_detection_rate_table_file)
  knitr::kable(goi_detection_rate_table)
}
```

```{r}
if(file.exists(params$plot_list_qc_probes_file)) {
  plot_list_qc_probes <- readRDS(params$plot_list_qc_probes_file)
  for(item in plot_list_qc_probes) {
    suppressMessages(capture.output(print(item)))
  }
}

```

# Normalization
```{r fig.width=12, fig.height=8}
if(file.exists(params$plot_list_normalization_file)) {
  plot_list_normalization <- readRDS(params$plot_list_normalization_file)
  for(item in plot_list_normalization) {
    suppressMessages(capture.output(print(item)))
  }
}
```

# 16S analysis
```{r}
if(file.exists(params$plot_list_16s_file)) {
  plot_list_16s <- readRDS(params$plot_list_16s_file)
  for(item in plot_list_16s) {
    suppressMessages(capture.output(print(item)))
  }
}
```

# Unsupervised analysis
## Dimensionality reduction
```{r fig.width=12, fig.height=12}
if(file.exists(params$plot_list_unsupervised_clustering_grid_file)) {
  plot_list_unsupervised_clustering_grids <- readRDS(params$plot_list_unsupervised_clustering_grid_file)
  for(item in plot_list_unsupervised_clustering_grids) {
    suppressMessages(capture.output(print(item)))
  }
}
```

## 16S score
```{r}
if(file.exists(params$plot_list_16s_score_file)) {
  plot_list_16s_scores <- readRDS(params$plot_list_16s_score_file)
  for(item in plot_list_16s_scores) {
    suppressMessages(capture.output(print(item)))
  }
}
```

## Heatmap of high-CV genes
```{r fig.width=12, fig.height=12}
if(file.exists(params$plot_list_cv_heatmap_file)) {
  plot_list_cv_heatmaps <- readRDS(params$plot_list_cv_heatmap_file)
  for(item in plot_list_cv_heatmaps) {
    grid::grid.newpage()
    grid::grid.draw(item$gtable)  # pheatmap returns a list with a "gtable" component
  }
}
```

# Differential expression analysis
```{r fig.width=12, fig.height=12}
# # Recursive function to traverse and draw grobs
# draw_grobs <- function(x) {
#   if (is.list(x)) {
#     # If x is a list, recursively apply draw_grobs to each element
#     lapply(x, draw_grobs)
#   } else if (inherits(x, "grob")) {
#     # If x is a grob, draw it
#     grid.draw(x)
#   }
# }

if(file.exists(params$plot_list_diff_exprs_grid_file)) {
  plot_list_diff_exprs_grids <- readRDS(params$plot_list_diff_exprs_grid_file)
  # draw_grobs(plot_list_diff_exprs_grids)
  for(item in plot_list_diff_exprs_grids) {
    suppressMessages(capture.output(print(item)))
  }
}
```

# Pathway analysis
```{r fig.width=12, fig.height=12}
if(file.exists(params$plot_list_pathway_analysis_grid_file)) {
  plot_list_pathway_analysis_grids <- readRDS(params$plot_list_pathway_analysis_grid_file)
  for(item in plot_list_pathway_analysis_grids) {
    suppressMessages(capture.output(print(item)))
  }
}
```

# Immune deconvolution
```{r fig.width=12, fig.height=12}
if(file.exists(params$plot_list_imm_decon_file)) {
  plot_list_imm_decon <- readRDS(params$plot_list_imm_decon_file)
  for(item in plot_list_imm_decon) {
    suppressMessages(capture.output(print(item)))
  }
}
```

# TCR analysis
```{r fig.width=12, fig.height=12}
if(file.exists(params$plot_list_tcr_analysis_grid_file)) {
  plot_list_tcr_analysis <- readRDS(params$plot_list_tcr_analysis_grid_file)
  for(item in plot_list_tcr_analysis) {
    suppressMessages(capture.output(print(item)))
  }
}
```

# Session info
```{r}
sessionInfo()
```